p2p/swarm.go

```diff
@@ -944,22 +948,27 @@ func (s *swarm) getListeners(
 		if gateway != nil {
 			err := nat_traversal.AcquirePortFromGateway(gateway, uint16(port))
 			if err != nil {
-				if err := tcpListener.Close(); err != nil {
-					s.logger.With().Error("error closing tcp listener", log.Err(err))
-				}
-				if err := udpListener.Close(); err != nil {
-					s.logger.With().Error("error closing udp listener", log.Err(err))
+				if upnpFails >= UPNPRetries {
+					return tcpListener, udpListener, nil
 				}
 				if randomPort {
+					if err := tcpListener.Close(); err != nil {
+						s.logger.With().Error("error closing tcp listener", log.Err(err))
+					}
+					if err := udpListener.Close(); err != nil {
+						s.logger.With().Error("error closing udp listener", log.Err(err))
+					}
 					port = 0
+					upnpFails++
```

Comment:
don't we want to increase this even if the port isn't random ?

Reply:
if the port isn't random we just log a warning and we don't ever repeat the loop, so this doesn't matter.

Reply:
cool.